apps/web/utils/discord-markdown.ts

```diff
@@ -3,77 +3,158 @@ import { load } from 'cheerio'
+  ) || null
+
+}, [userId], { revalidate: QUERY_REVALIDATE_TIME })()
+
+const fetchChannel = (channelId: string) => unstable_cache(async () => {
+  return (
+    await db
+      .selectFrom('channels')
+      .select(['snowflakeId', 'name'])
+      .where('snowflakeId', '=', channelId)
+      .executeTakeFirst()
+  ) || null
+
+}, [channelId], { revalidate: QUERY_REVALIDATE_TIME })()
+
+const fetchPost = (postId: string) => unstable_cache(async () => {
+  return (
+    await db
+      .selectFrom('posts')
+      .select(['snowflakeId', 'title'])
+      .where('snowflakeId', '=', postId)
+      .executeTakeFirst()
+  ) || null
+
+}, [postId], { revalidate: QUERY_REVALIDATE_TIME })()
+
+const channelLinkRegex = /https:\/\/discord\.com\/channels\/(?<guild>\d+)\/(?<channel>\d+)(\/(?<message>\d+))?/g;
+const userMention = /<@!?(?<user>\d+)>/g;
+const channelMention = /<#(?<channel>\d+)>/g;
+
+export const extractMentions = (content: string) => {
+  const postIds = new Set<string>(Array.from(content.matchAll(channelLinkRegex), (m) => m.groups?.channel ?? ''))
+  const memberIds = new Set<string>(Array.from(content.matchAll(userMention), (m) => m.groups?.user ?? ''))
+  const channelIds = new Set<string>(Array.from(content.matchAll(channelMention), (m) => m.groups?.channel ?? ''))
+  return { postIds, memberIds, channelIds }
+}
+
+export const fetchMentions = async (content: string) => {
+  const { postIds, memberIds, channelIds } = extractMentions(content)
+
+  // Fetch from db/cache
+  const [posts, users, channels] = (await Promise.all([
+    Promise.all(Array.from(postIds).map(fetchPost)),
+    Promise.all(Array.from(memberIds).map(fetchUser)),
+    Promise.all(Array.from(channelIds).map(fetchChannel)),
+  ]))
+
+  // Filter out null values
+  const postsFiltered = posts.filter((p) => p) as PostCache[]
+  const usersFiltered = users.filter((u) => u) as UserCache[]
+  const channelsFiltered = channels.filter((c) => c) as ChannelCache[]
+
+  return {
+    posts: postsFiltered,
+    users: usersFiltered,
+    channels: channelsFiltered
+  }
+}
+
+const internalLink = (content: string, posts: PostCache[]) => {
+  // Replace internal links
+  content = content.replace(channelLinkRegex, (match, guildId, channelId, _, messageId) => {
+    const post = posts.find((p) => p.snowflakeId === channelId)
     if (!post) return match
-    return `${getCanonicalPostUrl(post.snowflakeId)}${message ? `#message-${message}` : ''}`
+    return `${getCanonicalPostUrl(post.snowflakeId)}${messageId ? `#message-${messageId}` : ''}`
   })
+  return content
+}
 
-  // The library doesn't allow async callbacks, so we have to do this in two steps
-  toHTML(content, {
-    discordOnly: true,
-    discordCallback: {
-      user(node) {
-        memberIds.add(node.id)
-        return ''
-      },
-      channel(node) {
-        channelIds.add(node.id)
-        return ''
-      },
-    },
+export const parseDiscordMessageBasic = async (content: string) => {
```

Comment:
Can't we use `discord-markdown` here as well? Otherwise we will still have markdown stuff in the text like `**this**`. We can do a pass to convert to HTML then maybe call `.text()` from cheerio to extract the plain text from the document

Reply:
I just assumed it was inefficient to do more parsing then necessary and then re-parse the HTML into test. As well as Google __could__ understand the markdown and simply ignore. This could however be good for links because the simple won't have the long link (with masked links) but a simple text. I think possibly using a library to remove [markdown syntax](https://www.npmjs.com/package/remove-markdown) might be a better solution even though it gets back the the issue of having the Discord mentions twice. 

But I'll do this as it will simplify the code base (don't need to over complicate every thing).